From 80704d84055d8f33cd66824d78d16b89fc45db45 Mon Sep 17 00:00:00 2001
From: Emmanuele Bassi <ebassi@gnome.org>
Date: Wed, 1 Apr 2020 18:11:55 +0100
Subject: [PATCH] Check the memset length argument

Avoid overflows by using the checked multiplication macro for gsize.

Fixes: #132
---
 gdk-pixbuf/io-gif-animation.c | 6 +++++-
 1 file changed, 5 insertions(+), 1 deletion(-)

diff --git a/gdk-pixbuf/io-gif-animation.c b/gdk-pixbuf/io-gif-animation.c
index c9db3c66e..49674fd2e 100644
--- a/gdk-pixbuf/io-gif-animation.c
+++ b/gdk-pixbuf/io-gif-animation.c
@@ -412,11 +412,15 @@ gdk_pixbuf_gif_anim_iter_get_pixbuf (GdkPixbufAnimationIter *anim_iter)
 
         /* If no rendered frame, render the first frame */
         if (anim->last_frame == NULL) {
+                gsize len = 0;
                 if (anim->last_frame_data == NULL)
                         anim->last_frame_data = gdk_pixbuf_new (GDK_COLORSPACE_RGB, TRUE, 8, anim->width, anim->height);
                 if (anim->last_frame_data == NULL)
                         return NULL;
-                memset (gdk_pixbuf_get_pixels (anim->last_frame_data), 0, gdk_pixbuf_get_rowstride (anim->last_frame_data) * anim->height);
+                if (g_size_checked_mul (&len, gdk_pixbuf_get_rowstride (anim->last_frame_data), anim->height))
+                        memset (gdk_pixbuf_get_pixels (anim->last_frame_data), 0, len);
+                else
+                        return NULL;
                 composite_frame (anim, g_list_nth_data (anim->frames, 0));
         }
 
-- 
GitLab

